onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc

```diff
@@ -19,24 +21,42 @@ using android::nn::wrapper::OperandType, android::nn::wrapper::Type;
 
 Status AddNnapiTranspose(ModelBuilder& model_builder,
                          const std::string& data_input,
-                         const std::string& perm_input, const std::vector<int32_t>& perm,
-                         const std::string& output) {
-  auto& shaper(model_builder.GetShaper());
+                         const std::string& perm_input,
+                         const gsl::span<const int32_t> perm,
+                         const std::string& output,
+                         const Shape* output_shape) {
+  if (output_shape == nullptr) {
+    auto& shaper(model_builder.GetShaper());
+
+    // Calculate transpose output shape
+    {
+      const Shape& input_dimen = shaper[data_input];
+      ORT_RETURN_IF_NOT(perm.size() == input_dimen.size(), "Invalid perm is given!");
+      size_t size = input_dimen.size();
+      Shape output_dimen(size);
+      for (size_t i = 0; i < size; i++)
+        output_dimen[i] = input_dimen[perm[i]];
+
+      shaper.AddShape(output, output_dimen);
+    }
+    const auto shape = shaper[output];
+    output_shape = &shape;
+  }
```

Comment:
You're setting a pointer to the address of a local variable which will almost immediately become invalid. 

If we need to support calling this with an empty output_shape (that seems unnecessary though - caller can ignore the output_shape) pass in std::optional<Shape>&. Otherwise pass in Shape& and assign it here (i.e. `output_shape = shaper[output];`